pkg/csi_driver/node.go

```diff
@@ -496,58 +497,56 @@ func getFSStat(path string) (available, capacity, used, inodesFree, inodes, inod
-		klog.Infof("NodeStageVolume creating configmap %+v with data %v for volume %s", klog.KObj(cm), data, volumeID)
+		// Update the configMap after successful nfs mount operation.
+		nodeName := s.driver.config.NodeName
+		configmapName := lockrelease.ConfigMapNamePrefix + nodeName
+		klog.Infof("NodeStageVolume getting configmap %s/%s for volume %s", util.ManagedFilestoreCSINamespace, configmapName, volumeID)
 		start := time.Now()
-		cm, err := s.lockReleaseController.CreateConfigMapWithData(ctx, configmapName, util.ManagedFilestoreCSINamespace, data)
+		cm, err := s.lockReleaseController.GetConfigMap(ctx, configmapName, util.ManagedFilestoreCSINamespace)
 		duration := time.Since(start)
-		s.lockReleaseController.RecordKubeAPIMetrics(err, metrics.ConfigMapResourceType, metrics.CreateOpType, metrics.NodeStageOpSource, duration)
+		s.lockReleaseController.RecordKubeAPIMetrics(err, metrics.ConfigMapResourceType, metrics.GetOpType, metrics.NodeStageOpSource, duration)
 		if err != nil {
-			klog.Errorf("NodeStageVolume failed to create configmap %+v with data %s for volume %s: %v", klog.KObj(cm), data, volumeID, err)
+			klog.Errorf("NodeStageVolume failed to get configmap %s/%s for volume %s: %v", util.ManagedFilestoreCSINamespace, configmapName, volumeID, err)
 			return err
 		}
-		klog.Infof("NodeStageVolume successfully created configmap %+v with data %v for volume %s", klog.KObj(cm), cm.Data, volumeID)
-		return nil
-	}
 
-	if err := s.lockReleaseController.UpdateConfigMapWithKeyValue(ctx, cm, lockInfoKey, filestoreIP); err != nil {
-		klog.Errorf("NodeStageVolume failed to update configmap %+v with lock info {%s: %s} for volume %s: %v", klog.KObj(cm), lockInfoKey, filestoreIP, volumeID, err)
-		return err
-	}
+		lockInfoKey, err := s.generateLockInfoKeyFromVolumeID(volumeID)
+		if err != nil {
+			klog.Errorf("NodeStageVolume failed to generate lock info key for volume %s: %v", volumeID, err)
+			return err
+		}
+
+		// Create or update the configmap with lock info.
+		filestoreIP := attr[attrIP]
+		if cm == nil {
+			data := map[string]string{lockInfoKey: filestoreIP}
+			klog.Infof("NodeStageVolume creating configmap %+v with data %v for volume %s", klog.KObj(cm), data, volumeID)
+			start := time.Now()
+			cm, err := s.lockReleaseController.CreateConfigMapWithData(ctx, configmapName, util.ManagedFilestoreCSINamespace, data)
+			duration := time.Since(start)
+			s.lockReleaseController.RecordKubeAPIMetrics(err, metrics.ConfigMapResourceType, metrics.CreateOpType, metrics.NodeStageOpSource, duration)
+			if err != nil {
+				klog.Errorf("NodeStageVolume failed to create configmap %+v with data %s for volume %s: %v", klog.KObj(cm), data, volumeID, err)
+				return err
+			}
+			klog.Infof("NodeStageVolume successfully created configmap %+v with data %v for volume %s", klog.KObj(cm), cm.Data, volumeID)
+			return nil
+		}
+
+		updateErr := s.lockReleaseController.UpdateConfigMapWithKeyValue(ctx, cm, lockInfoKey, filestoreIP)
```

Comment:
I think this is the only place (UPDATE configmap) we need to have a retry mechanism, rather than the entire node stage volume function.